86.md

```diff
@@ -0,0 +1,88 @@
+
+```json
+{
+  "method": "<method-name>",
+  "params": ["<array>", "<of>", "<parameters>"]
+}
+```
+
+Then it should return a response in the format
+
+```json
+{
+  "result": {"<arbitrary>": "<value>"},
+  "error": "<optional error message, if the call has errored>"
+}
+```
+
+This is the list of **methods** that may be supported:
+
+* `supportedmethods`:
+  - params: `[]`
+  - result: `["<method-name>", "<method-name>", ...]` (an array with the names of all the other supported methods)
+* `banpubkey`:
+  - params: `["<32-byte-hex-public-key>", "<optional-reason>"]`
+  - result: `true` (a boolean always set to `true`)
+* `listbannedpubkeys`:
+  - params: `[]`
+  - result: `[{"pubkey": "<32-byte-hex>", "reason": "<optional-reason>"}, ...]`, an array of objects
+* `allowpubkey`:
+  - params: `["<32-byte-hex-public-key>", "<optional-reason>"]`
+  - result: `true` (a boolean always set to `true`)
+* `listallowedpubkeys`:
+  - params: `[]`
+  - result: `[{"pubkey": "<32-byte-hex>", "reason": "<optional-reason>"}, ...]`, an array of objects
+* `listeventsneedingmoderation`:
+  - params: `[]`
+  - result: `[{"id": "<32-byte-hex>", "reason": "<optional-reason>"}]`, an array of objects
+* `allowevent`:
+  - params: `["<32-byte-hex-event-id>", "<optional-reason>"]`
+  - result: `true` (a boolean always set to `true`)
+* `banevent`:
+  - params: `["<32-byte-hex-event-id>", "<optional-reason>"]`
+  - result: `true` (a boolean always set to `true`)
+* `listbannedevents`:
+  - params: `[]`
+  - result: `[{"id": "<32-byte hex>", "reason": "<optional-reason>"}, ...]`, an array of objects
+* `changerelayname`:
+  - params: `["<new-name>"]`
+  - result: `true` (a boolean always set to `true`)
+* `changerelaydescription`:
+  - params: `["<new-description>"]`
+  - result: `true` (a boolean always set to `true`)
+* `changerelayicon`:
+  - params: `["<new-icon-url>"]`
+  - result: `true` (a boolean always set to `true`)
+* `allowkind`:
+  - params: `[<kind-number>]`
+  - result: `true` (a boolean always set to `true`)
+* `disallowkind`:
+  - params: `[<kind-number>]`
+  - result: `true` (a boolean always set to `true`)
+* `listallowedkinds`:
+  - params: `[]`
+  - result: `[<kind-number>, ...]`, an array of numbers
+* `blockip`:
+  - params: `["<ip-address>", "<optional-reason>"]`
+  - result: `true` (a boolean always set to `true`)
+* `unblockip`:
+  - params: `["<ip-address>"]`
+  - result: `true` (a boolean always set to `true`)
+* `listblockedips`:
+  - params: `[]`
+  - result: `[{"ip": "<ip-address>", "reason": "<optional-reason>"}, ...]`, an array of objects
+
+### Authorization
+
+The request must contain an `Authorization` header with the value set to `Nostr <base64(auth-event)>` in which `<auth-event>` is an event of kind `27235` with a tag `["host", <hostname-of-target-relay>]`, a tag `["payload", <sha256-hash-of-the-request-body-as-lowercase-hex>]` and `created_at` set to the current timestamp.
```

Comment:
```suggestion
The request must contain an `Authorization` header with a valid [NIP-98](./98.md) event, except the `payload` tag is required. The `u` tag is the relay URL.
```

It's better to just link to NIP-98 for full context, including the nuance around the created_at timestamp since there might be a slight delay between signing the event and sending the request.

The `host` tag introduced here violates the principle of having one way to do things, and makes existing NIP-98 tooling not work. It should use the `u` tag with the full relay URL.

Reply:
Is there any NIP-98 tooling? It's garbage, let's burn it.

Reply:
Bro. :sob: I painstakingly implemented nip98 in Nostrify with a ton of tests: https://gitlab.com/soapbox-pub/nostrify/-/blob/main/src/NIP98.test.ts?ref_type=heads

Ditto uses it extensively in its HTTP API: https://gitlab.com/soapbox-pub/ditto/-/blob/main/src/middleware/auth98Middleware.ts?ref_type=heads

I was already mad that Blossom made me implement an entirely new HTTP authorization scheme. Aren't you the guy who doesn't want nip46 to change? nip98 is already perfect as it is. This `host` tag doesn't even make sense because it falls into the same pitfall of assuming that a relay URL is a host, when mine are at `/relay`

Reply:
I feel your pain, my point is just that NIP-98 offers no advantages to anyone.

The comparison with NIP-46 is unwarranted because there are no generic NIP-98-enabled clients talking to NIP-98-enabled servers -- it's always a proprietary, domain-specific API, and that API can use any authorization scheme it wants, there are no advantages to using NIP-98. In fact there are disadvantages.

For example, having the full URL is a burden because the server often doesn't know if it's under "http" or "https" and that's very annoying. Also in most cases the domain is enough, doesn't matter if your relay is under `/relay`, you still control the domain, so it's fine.

Reply:
It took me 20 lines in Go to make the damn `Authorization` header described in this NIP, but if you think that reusing a library is worth it then sure, I'm fine with it.

Reply:
This is a security feature. It's something that you absolutely want to use a library function for so it's tested and ideally audited because it's shared by multiple people.

Reply:
https://github.com/fiatjaf/nak/commit/809865ca0ccfd28b82b8c3f038b5891d4c58e81f

Reply:
https://github.com/fiatjaf/khatru/commit/214b92ba3b1ebd978fce4860a6e2ec6a096dea5a

Reply:
Thank you!